src/main/java/org/kohsuke/stapler/idea/extension/AbstractStructureViewExtension.java

```diff
@@ -0,0 +1,49 @@
+package org.kohsuke.stapler.idea.extension;
+
+import com.intellij.ide.highlighter.JavaFileType;
+import com.intellij.ide.structureView.StructureViewExtension;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.util.text.StringUtilRt;
+import com.intellij.psi.PsiElement;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Abstract class for jelly/java structure views
+ *
+ * @author Julien Greffe
+ */
+public abstract class AbstractStructureViewExtension implements StructureViewExtension {
+
+    private final static String SRC_JAVA = "src/main/java";
+    private final static String SRC_RESOURCES = "src/main/resources";
+
+    /**
+     * Replace src/main/java to src/main/resources (location of expected related jelly files)
+     */
+    protected String getExpectedJellyParentPath(String path) {
+        return path.replace(SRC_JAVA, SRC_RESOURCES);
+    }
+
+    protected String getExpectedJavaPath(String path) {
+        return path.replace(SRC_RESOURCES, SRC_JAVA);
+    }
+
+    /**
+     * Construct path without extension
+     */
+    protected CharSequence getPathWithoutExtension(PsiElement javaFile) {
+        CharSequence path = javaFile.getContainingFile().getVirtualFile().getPath();
+        int i = StringUtilRt.lastIndexOf(path, '.', 0, path.length());
+        return i < 0 ? path : path.subSequence(0, i);
+    }
```

Comment:
Assuming we are adding extension `com.intellij.psi.PsiClass`, this needs to be re-written. The current implementation doesn't handle inner classes like Descriptors.

Reply:
@duemir : could you detail please?

Reply:
Here we retrieve path from containing file `javaFile.getContainingFile()`, so I guess it works for inner classes (?)

Reply:
https://github.com/jenkinsci/design-library-plugin/blob/master/src/main/java/io/jenkins/plugins/designlibrary/HeteroList.java is a good test case.

It has two views, `configure.jelly` and `index.jelly` and it has 6 inner classes only four of which have their own views named `config.jelly`.

At the moment the structure view shows the same `index.jelly` and `configure.jelly` views for every single inner class even those that do not have views:

<img width="259" alt="Screen Shot 2022-10-11 at 15 37 10" src="https://user-images.githubusercontent.com/348580/195005378-e317cc6f-65f2-4e67-bcf0-58f201dff193.png">

Reply:
Ok, got your point! I'll have a look. 👍🏼

Reply:
Pushed again.